## Backtesting NVDA Options via Monte Carlo & Statistical Arbitrage on Historical Data

In the project we will perform Monte Carlo simulations on historical NVDA options data to see if we can make a profit on mispriced options. 

Some assumptions in this algorithm:

1. No new options are added
2. We only buy the options since this is historical data
3. The stock prices follow the black scholes model 

# NOTE: THIS IS NOT FINANCIAL ADVICE. I AM NOT A FINANCIAL ADVISOR. 

### 1. Data cleaning + transformations




Import libraries


In [1]:
import pandas as pd #linalg
import numpy as np #math 
from scipy.stats import norm #stats for mc sim
from datetime import datetime #datetime library 


import CBOE options data

In [2]:
jan_data = pd.read_csv("CBOE Data/NVDA/UnderlyingOptionsEODCalcs_2023-01.csv")
feb_data = pd.read_csv("CBOE Data/NVDA/UnderlyingOptionsEODCalcs_2023-02.csv")
mar_data = pd.read_csv("CBOE Data/NVDA/UnderlyingOptionsEODCalcs_2023-03.csv")
apr_data = pd.read_csv("CBOE Data/NVDA/UnderlyingOptionsEODCalcs_2023-04.csv")
may_data = pd.read_csv("CBOE Data/NVDA/UnderlyingOptionsEODCalcs_2023-05.csv")
jun_data = pd.read_csv("CBOE Data/NVDA/UnderlyingOptionsEODCalcs_2023-06.csv")
jul_data = pd.read_csv("CBOE Data/NVDA/UnderlyingOptionsEODCalcs_2023-07.csv")
aug_data = pd.read_csv("CBOE Data/NVDA/UnderlyingOptionsEODCalcs_2023-08.csv")
sep_data = pd.read_csv("CBOE Data/NVDA/UnderlyingOptionsEODCalcs_2023-09.csv")
oct_data = pd.read_csv("CBOE Data/NVDA/UnderlyingOptionsEODCalcs_2023-10.csv")
nov_data = pd.read_csv("CBOE Data/NVDA/UnderlyingOptionsEODCalcs_2023-11.csv")
dec_data = pd.read_csv("CBOE Data/NVDA/UnderlyingOptionsEODCalcs_2023-12.csv")

ls = [jan_data, feb_data, mar_data, apr_data, may_data, jun_data, jul_data, aug_data, sep_data, oct_data, nov_data, dec_data]
data = pd.concat(ls, ignore_index=True)


In [3]:
data.head()

underlying_symbol  quote_date  root  expiration  strike option_type   open  \
0              NVDA  2023-01-03  NVDA  2023-01-06    65.0           C  76.85   
1              NVDA  2023-01-03  NVDA  2023-01-06    65.0           P   0.00   
2              NVDA  2023-01-03  NVDA  2023-01-06    70.0           C  78.40   
3              NVDA  2023-01-03  NVDA  2023-01-06    70.0           P   0.01   
4              NVDA  2023-01-03  NVDA  2023-01-06    75.0           C   0.00   

    high    low  close  ...  rho_1545  bid_size_eod  bid_eod  ask_size_eod  \
0  76.90  76.85  76.90  ...    0.0060            25    77.35            25   
1   0.00   0.00   0.00  ...   -0.0000             0     0.00             6   
2  78.40  78.35  78.35  ...    0.0063            25    72.35            25   
3   0.01   0.01   0.01  ...   -0.0000             0     0.00             6   
4   0.00   0.00   0.00  ...    0.0067            25    67.35            25   

   ask_eod  underlying_bid_eod  underlying_ask_eod    vwap  open_interest  \
0    79.00              143.15              143.16  76.875              2   
1     0.01              143.15              143.16   0.000             36   
2    74.05              143.15              143.16  78.375              2   
3     0.01              143.15              143.16   0.010              2   
4    69.05              143.15              143.16   0.000             40   

   delivery_code  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  

[5 rows x 34 columns]

options have no id, define function to generate id based on symbol, expdate, strike price, and type

In [5]:
def generate_option_id(option):
    return f"{option['underlying_symbol']}_{option['expiration']}_{option['strike']}_{option['option_type']}"

In [6]:
option_id_series = data.apply(lambda row: generate_option_id(row), axis=1)
data.insert(0, 'option_id', option_id_series)

In [7]:
type(data['expiration'][0])

str

include an interest rate column

In [8]:
data.insert(data.shape[1], 'interest_rate',"")

In [9]:
data.head()

option_id underlying_symbol  quote_date  root  expiration  \
0  NVDA_2023-01-06_65.0_C              NVDA  2023-01-03  NVDA  2023-01-06   
1  NVDA_2023-01-06_65.0_P              NVDA  2023-01-03  NVDA  2023-01-06   
2  NVDA_2023-01-06_70.0_C              NVDA  2023-01-03  NVDA  2023-01-06   
3  NVDA_2023-01-06_70.0_P              NVDA  2023-01-03  NVDA  2023-01-06   
4  NVDA_2023-01-06_75.0_C              NVDA  2023-01-03  NVDA  2023-01-06   

   strike option_type   open   high    low  ...  bid_size_eod  bid_eod  \
0    65.0           C  76.85  76.90  76.85  ...            25    77.35   
1    65.0           P   0.00   0.00   0.00  ...             0     0.00   
2    70.0           C  78.40  78.40  78.35  ...            25    72.35   
3    70.0           P   0.01   0.01   0.01  ...             0     0.00   
4    75.0           C   0.00   0.00   0.00  ...            25    67.35   

   ask_size_eod  ask_eod  underlying_bid_eod  underlying_ask_eod    vwap  \
0            25    79.00              143.15              143.16  76.875   
1             6     0.01              143.15              143.16   0.000   
2            25    74.05              143.15              143.16  78.375   
3             6     0.01              143.15              143.16   0.010   
4            25    69.05              143.15              143.16   0.000   

   open_interest  delivery_code  interest_rate  
0              2            NaN                 
1             36            NaN                 
2              2            NaN                 
3              2            NaN                 
4             40            NaN                 

[5 rows x 36 columns]

finding interest rate data

In [10]:
import requests
import fredapi as fd

In [12]:
api_key = 'xxxxx'
url = 'https://api.stlouisfed.org/fred/category?category_id=125&api_key=' + api_key +'&file_type=json'
response = requests.get(url).json()
response

{'categories': [{'id': 125, 'name': 'Trade Balance', 'parent_id': 13}]}

In [13]:
fred = fd.Fred(api_key = api_key)
rate_data = fred.search('Federal Funds Effective Rate')
rate_data.head()

id realtime_start realtime_end  \
series id                                          
FEDFUNDS    FEDFUNDS     2024-08-13   2024-08-13   
DFF              DFF     2024-08-13   2024-08-13   
FF                FF     2024-08-13   2024-08-13   
RIFSPFFNB  RIFSPFFNB     2024-08-13   2024-08-13   
RIFSPFFNA  RIFSPFFNA     2024-08-13   2024-08-13   

                                  title observation_start observation_end  \
series id                                                                   
FEDFUNDS   Federal Funds Effective Rate        1954-07-01      2024-07-01   
DFF        Federal Funds Effective Rate        1954-07-01      2024-08-12   
FF         Federal Funds Effective Rate        1954-07-07      2024-08-07   
RIFSPFFNB  Federal Funds Effective Rate        1954-07-01      2024-08-12   
RIFSPFFNA  Federal Funds Effective Rate        1955-01-01      2023-01-01   

                          frequency frequency_short    units units_short  \
series id                                                                  
FEDFUNDS                    Monthly               M  Percent           %   
DFF                    Daily, 7-Day               D  Percent           %   
FF         Weekly, Ending Wednesday               W  Percent           %   
RIFSPFFNB                     Daily               D  Percent           %   
RIFSPFFNA                    Annual               A  Percent           %   

               seasonal_adjustment seasonal_adjustment_short  \
series id                                                      
FEDFUNDS   Not Seasonally Adjusted                       NSA   
DFF        Not Seasonally Adjusted                       NSA   
FF         Not Seasonally Adjusted                       NSA   
RIFSPFFNB  Not Seasonally Adjusted                       NSA   
RIFSPFFNA  Not Seasonally Adjusted                       NSA   

                        last_updated popularity  \
series id                                         
FEDFUNDS   2024-08-01 15:17:51-05:00         98   
DFF        2024-08-13 15:17:42-05:00         86   
FF         2024-08-08 15:16:46-05:00         63   
RIFSPFFNB  2024-08-13 15:17:38-05:00         44   
RIFSPFFNA  2024-01-02 15:26:02-06:00         42   

                                                       notes  
series id                                                     
FEDFUNDS   Averages of daily figures.   For additional hi...  
DFF        For additional historical federal funds rate d...  
FF         Averages of daily figures.   For additional hi...  
RIFSPFFNB  For additional historical federal funds rate d...  
RIFSPFFNA  Averages of daily figures.   For additional hi...

We use the DFF series which contains the daily effective rates

In [14]:
rate = fred.get_series('DFF').to_frame()

In [15]:
rate

0
1954-07-01  1.13
1954-07-02  1.25
1954-07-03  1.25
1954-07-04  1.25
1954-07-05  0.88
...          ...
2024-08-08  5.33
2024-08-09  5.33
2024-08-10  5.33
2024-08-11  5.33
2024-08-12  5.33

[25611 rows x 1 columns]

In [16]:
year_rate = rate['2023-01-03':'2023-12-29']
year_rate = year_rate.reset_index()
year_rate.rename(columns={'index': 'date'}, inplace=True)
year_rate.rename(columns={0: 'interest_rate'}, inplace=True)
year_rate

date  interest_rate
0   2023-01-03           4.33
1   2023-01-04           4.33
2   2023-01-05           4.33
3   2023-01-06           4.33
4   2023-01-07           4.33
..         ...            ...
356 2023-12-25           5.33
357 2023-12-26           5.33
358 2023-12-27           5.33
359 2023-12-28           5.33
360 2023-12-29           5.33

[361 rows x 2 columns]

In [17]:
year_rate['date'] = pd.to_datetime(year_rate['date']).dt.date
data['quote_date'] = pd.to_datetime(data['quote_date']).dt.date

In [18]:
year_rate.set_index('date', inplace=True)

In [19]:
data['interest_rate'] = data['quote_date'].map(year_rate['interest_rate'])

In [20]:
data

option_id underlying_symbol  quote_date  root  \
0         NVDA_2023-01-06_65.0_C              NVDA  2023-01-03  NVDA   
1         NVDA_2023-01-06_65.0_P              NVDA  2023-01-03  NVDA   
2         NVDA_2023-01-06_70.0_C              NVDA  2023-01-03  NVDA   
3         NVDA_2023-01-06_70.0_P              NVDA  2023-01-03  NVDA   
4         NVDA_2023-01-06_75.0_C              NVDA  2023-01-03  NVDA   
...                          ...               ...         ...   ...   
1040997  NVDA_2026-06-18_910.0_P              NVDA  2023-12-29  NVDA   
1040998  NVDA_2026-06-18_920.0_C              NVDA  2023-12-29  NVDA   
1040999  NVDA_2026-06-18_920.0_P              NVDA  2023-12-29  NVDA   
1041000  NVDA_2026-06-18_930.0_C              NVDA  2023-12-29  NVDA   
1041001  NVDA_2026-06-18_930.0_P              NVDA  2023-12-29  NVDA   

         expiration  strike option_type   open   high    low  ...  \
0        2023-01-06    65.0           C  76.85  76.90  76.85  ...   
1        2023-01-06    65.0           P   0.00   0.00   0.00  ...   
2        2023-01-06    70.0           C  78.40  78.40  78.35  ...   
3        2023-01-06    70.0           P   0.01   0.01   0.01  ...   
4        2023-01-06    75.0           C   0.00   0.00   0.00  ...   
...             ...     ...         ...    ...    ...    ...  ...   
1040997  2026-06-18   910.0           P   0.00   0.00   0.00  ...   
1040998  2026-06-18   920.0           C  45.33  45.48  45.33  ...   
1040999  2026-06-18   920.0           P   0.00   0.00   0.00  ...   
1041000  2026-06-18   930.0           C  44.40  46.21  43.43  ...   
1041001  2026-06-18   930.0           P   0.00   0.00   0.00  ...   

         bid_size_eod  bid_eod  ask_size_eod  ask_eod  underlying_bid_eod  \
0                  25    77.35            25    79.00              143.15   
1                   0     0.00             6     0.01              143.15   
2                  25    72.35            25    74.05              143.15   
3                   0     0.00             6     0.01              143.15   
4                  25    67.35            25    69.05              143.15   
...               ...      ...           ...      ...                 ...   
1040997            23   412.00            23   419.90              495.17   
1040998           184    45.55            60    47.20              495.17   
1040999            30   421.35            22   429.30              495.17   
1041000           219    44.25           140    46.15              495.17   
1041001            25   428.90            25   442.65              495.17   

         underlying_ask_eod     vwap  open_interest  delivery_code  \
0                    143.16  76.8750              2            NaN   
1                    143.16   0.0000             36            NaN   
2                    143.16  78.3750              2            NaN   
3                    143.16   0.0100              2            NaN   
4                    143.16   0.0000             40            NaN   
...                     ...      ...            ...            ...   
1040997              495.26   0.0000              0            NaN   
1040998              495.26  45.4529             35            NaN   
1040999              495.26   0.0000              0            NaN   
1041000              495.26  44.5626             38            NaN   
1041001              495.26   0.0000              0            NaN   

         interest_rate  
0                 4.33  
1                 4.33  
2                 4.33  
3                 4.33  
4                 4.33  
...                ...  
1040997           5.33  
1040998           5.33  
1040999           5.33  
1041000           5.33  
1041001           5.33  

[1041002 rows x 36 columns]

In [21]:
data.to_csv('options_data.csv', index=False)

Run cell below for "data" dataframe without running lines above

In [22]:
data = pd.read_csv("options_data.csv")

In [23]:
data

option_id underlying_symbol  quote_date  root  \
0         NVDA_2023-01-06_65.0_C              NVDA  2023-01-03  NVDA   
1         NVDA_2023-01-06_65.0_P              NVDA  2023-01-03  NVDA   
2         NVDA_2023-01-06_70.0_C              NVDA  2023-01-03  NVDA   
3         NVDA_2023-01-06_70.0_P              NVDA  2023-01-03  NVDA   
4         NVDA_2023-01-06_75.0_C              NVDA  2023-01-03  NVDA   
...                          ...               ...         ...   ...   
1040997  NVDA_2026-06-18_910.0_P              NVDA  2023-12-29  NVDA   
1040998  NVDA_2026-06-18_920.0_C              NVDA  2023-12-29  NVDA   
1040999  NVDA_2026-06-18_920.0_P              NVDA  2023-12-29  NVDA   
1041000  NVDA_2026-06-18_930.0_C              NVDA  2023-12-29  NVDA   
1041001  NVDA_2026-06-18_930.0_P              NVDA  2023-12-29  NVDA   

         expiration  strike option_type   open   high    low  ...  \
0        2023-01-06    65.0           C  76.85  76.90  76.85  ...   
1        2023-01-06    65.0           P   0.00   0.00   0.00  ...   
2        2023-01-06    70.0           C  78.40  78.40  78.35  ...   
3        2023-01-06    70.0           P   0.01   0.01   0.01  ...   
4        2023-01-06    75.0           C   0.00   0.00   0.00  ...   
...             ...     ...         ...    ...    ...    ...  ...   
1040997  2026-06-18   910.0           P   0.00   0.00   0.00  ...   
1040998  2026-06-18   920.0           C  45.33  45.48  45.33  ...   
1040999  2026-06-18   920.0           P   0.00   0.00   0.00  ...   
1041000  2026-06-18   930.0           C  44.40  46.21  43.43  ...   
1041001  2026-06-18   930.0           P   0.00   0.00   0.00  ...   

         bid_size_eod  bid_eod  ask_size_eod  ask_eod  underlying_bid_eod  \
0                  25    77.35            25    79.00              143.15   
1                   0     0.00             6     0.01              143.15   
2                  25    72.35            25    74.05              143.15   
3                   0     0.00             6     0.01              143.15   
4                  25    67.35            25    69.05              143.15   
...               ...      ...           ...      ...                 ...   
1040997            23   412.00            23   419.90              495.17   
1040998           184    45.55            60    47.20              495.17   
1040999            30   421.35            22   429.30              495.17   
1041000           219    44.25           140    46.15              495.17   
1041001            25   428.90            25   442.65              495.17   

         underlying_ask_eod     vwap  open_interest  delivery_code  \
0                    143.16  76.8750              2            NaN   
1                    143.16   0.0000             36            NaN   
2                    143.16  78.3750              2            NaN   
3                    143.16   0.0100              2            NaN   
4                    143.16   0.0000             40            NaN   
...                     ...      ...            ...            ...   
1040997              495.26   0.0000              0            NaN   
1040998              495.26  45.4529             35            NaN   
1040999              495.26   0.0000              0            NaN   
1041000              495.26  44.5626             38            NaN   
1041001              495.26   0.0000              0            NaN   

         interest_rate  
0                 4.33  
1                 4.33  
2                 4.33  
3                 4.33  
4                 4.33  
...                ...  
1040997           5.33  
1040998           5.33  
1040999           5.33  
1041000           5.33  
1041001           5.33  

[1041002 rows x 36 columns]

In [24]:
data['quote_date'] = pd.to_datetime(data['quote_date'])

In [25]:
data['expiration'] = pd.to_datetime(data['expiration'])

## 2. Define Portfolio class

In [51]:
import pandas as pd
import numpy as np
from datetime import datetime

class Portfolio:
    def __init__(self, money):
        self.money = money
        self.positions = pd.DataFrame(columns=['option_id', 'underlying_symbol', 'option_type', 'quantity', 'strike', 'expiration', 'price', 'stock_price']) 
        self.owned_stocks = pd.DataFrame(columns=['symbol', 'quantity', 'type', 'stock_price'])
        self.transaction_log = pd.DataFrame(columns=['date', 'action', 'option_id', 'underlying_symbol', 'option_type', 'quantity', 'price', 'total_cost'])
        self.daily_pnl_log = pd.DataFrame(columns=['date', 'pnl'])

    def log_transaction(self, date, action, option_id, symbol, opt_type, quantity, price, total_cost):
        new_transaction = {
            'date': date,
            'action': action,
            'option_id': option_id,
            'underlying_symbol': symbol,
            'option_type': opt_type,
            'quantity': quantity,
            'price': price,
            'total_cost': total_cost
        }
        self.transaction_log = pd.concat([self.transaction_log, pd.DataFrame([new_transaction])], ignore_index=True)
    
    def log_daily_pnl(self, date):
        daily_pnl = self.calculate_daily_pnl()
        new_pnl = {'date': date, 'pnl': daily_pnl}
        self.daily_pnl_log = pd.concat([self.daily_pnl_log, pd.DataFrame([new_pnl])], ignore_index=True)

    def calculate_daily_pnl(self):
        # Calculate the P&L for all positions
        pnl = 0
        for idx, row in self.positions.iterrows():
            current_price = row['stock_price']
            if row['option_type'] == 'C':
                pnl += max(current_price - row['strike'], 0) * row['quantity']
            elif row['option_type'] == 'P':
                pnl += max(row['strike'] - current_price, 0) * row['quantity']
        return pnl

    def buy_option(self, data, optionid, symbol, opt_type, quantity, strike, expiration, price, stock_price):
        if self.money <= 0:
            return
        if quantity == 0 or price == 0:
            return
        try:
            quantity = int(quantity)
            price = float(price)
            cost = np.round((price * quantity), 3)
        except ValueError:
            print("invalid quantity or price.")
            return
        
        if self.money >= cost:
            new_position = pd.DataFrame({
                'option_id': [optionid],
                'underlying_symbol': [symbol],
                'option_type': [opt_type],
                'quantity': [quantity],
                'strike': [strike],
                'expiration': [expiration],
                'unit_price': [price],
                'total_cost': [cost],
                'stock_price': [stock_price]  
            })
            # Append newly purchased options to positions using pd.concat
            self.positions = pd.concat([self.positions, new_position], ignore_index=True)
            # Update capital
            self.money -= cost
            
            # Log the transaction
            self.log_transaction(data['quote_date'], 'BUY', optionid, symbol, opt_type, quantity, price, cost)
    
    def perform_monte_carlo_simulation(self, S0, K, T, r, sigma):
        """
        Perform Monte Carlo simulation for option pricing.
        """
        N = 1_000_000 
        # Calculate payoffs
        payoff = np.zeros(N)
        for i in range(N):
            WT = np.sqrt(T) * np.random.randn()
            ST = S0 * np.exp((r - 0.5 * sigma**2) * T + sigma * WT)
            payoff[i] = np.exp(-r * T) * max(ST - K, 0)
        
        MC_option_price = np.mean(payoff)
        MC_error = 3 * np.sqrt(np.var(payoff)) / np.sqrt(N)
        return MC_option_price, MC_error

    def solve_binomial_model(self, S0, K, T, r, sigma, N=5, opttype='C'):
        dt = T / N
        u = np.exp(sigma * np.sqrt(dt))
        d = 1 / u
        q = (np.exp(r * dt) - d) / (u - d)
        disc = np.exp(-r * dt)

        # Initialize stock prices at maturity
        S = S0 * d**(np.arange(N, -1, -1)) * u**(np.arange(0, N + 1, 1))

        # Option payoff at maturity
        if opttype == 'P':
            C = np.maximum(0, K - S)
        else:
            C = np.maximum(0, S - K)

        # Backward recursion through the tree
        for i in np.arange(N - 1, -1, -1):
            S = S0 * d**(np.arange(i, -1, -1)) * u**(np.arange(0, i + 1, 1))
            C[:i + 1] = disc * (q * C[1:i + 2] + (1 - q) * C[0:i + 1])

            # Return the option price today, and prices for up and down movements
            if i == N - 1:
                option_price_today = C[0]
                option_price_up = C[1]
                option_price_down = C[0]
                return option_price_today, option_price_up, option_price_down

            C = C[:-1]
            if opttype == 'P':
                C = np.maximum(C, K - S)
            else:
                C = np.maximum(C, S - K)

        return C[0]
    
    def solve_continuation_value(self, S0, K, T, r, sigma, N=10, opttype='C'):
        option_price_today, option_price_up, option_price_down = self.solve_binomial_model(S0, K, T, r, sigma, N, opttype)
        dt = T / N
        u = np.exp(sigma * np.sqrt(dt))
        d = 1 / u
        q_u = (np.exp(r * dt) - d) / (u - d)
        disc = np.exp(-r * dt)

        # Calculate continuation value
        continuation_value = disc * (q_u * option_price_up + (1 - q_u) * option_price_down)
        return continuation_value

    def exercise_option(self, current_date):
        for idx, row in self.positions.iterrows():
            S0 = row['stock_price']
            K = row['strike']
            T = (row['expiration'] - current_date).days / 365.0
            r = 0.01  # Assuming a fixed interest rate
            sigma = 0.2  # Assuming a fixed volatility

            # Ensure the minimum holding period has passed
            if (current_date - row['purchase_date']) < self.min_holding_period:
                continue

            # Get the current value and continuation value using the binomial model
            current_value, _, _ = self.solve_binomial_model(S0, K, T, r, sigma, opttype=row['option_type'])
            continuation_value = self.solve_continuation_value(S0, K, T, r, sigma, opttype=row['option_type'])

            # Only exercise if the immediate exercise value is greater than the continuation value
            if current_value > continuation_value:
                self.settle_option(row['option_id'], row['underlying_symbol'], row['option_type'], row['quantity'], row['strike'], row['expiration'], row['unit_price'])

            
    def settle_option(self, option_id, symbol, opt_type, quantity, strike, expiration, price):
        S0 = self.positions[self.positions['option_id'] == option_id]['stock_price'].iloc[0]
        profit = 0
        
        if opt_type == 'C':
            profit = (S0 - strike) * quantity
            self.money += profit
        
        elif opt_type == 'P':
            profit = (strike - S0) * quantity
            self.money += profit

        # Log the transaction as a sale
        self.log_transaction(expiration, 'SELL', option_id, symbol, opt_type, quantity, price, profit)
        
        # Remove the exercised option from the portfolio
        self.positions = self.positions[self.positions['option_id'] != option_id]

    def hedge_option(self, S0, K, T, r, sigma, quantity):
        u = np.exp(sigma * np.sqrt(T))
        d = 1 / u
        _, C_u, C_d = self.solve_binomial_model(S0, K, T, r, sigma)

        # Calculate stock prices after an up and down movement
        S_u = S0 * u
        S_d = S0 * d

        # Calculate delta
        delta = (C_u - C_d) / (S_u - S_d)
        stocks_to_hedge = delta * quantity

        if delta < 0:
            # Short stock
            self.money += -stocks_to_hedge * S0
            self.owned_stocks = pd.concat([self.owned_stocks, pd.DataFrame([{'symbol': 'S0', 'quantity': -stocks_to_hedge, 'type': 'short', 'stock_price': S0}])], ignore_index=True)
        elif delta > 0:
            # Buy stock
            self.money -= stocks_to_hedge * S0
            self.owned_stocks = pd.concat([self.owned_stocks, pd.DataFrame([{'symbol': 'S0', 'quantity': stocks_to_hedge, 'type': 'long', 'stock_price': S0}])], ignore_index=True)

    def close_option(self, option_id, symbol, opt_type, quantity, strike, expiration, price):
        S0 = self.positions[self.positions['option_id'] == option_id]['stock_price'].iloc[0]
        
        if opt_type == 'C' and S0 > strike:
            self.settle_option(option_id, symbol, opt_type, quantity, strike, expiration, price)
            print(f"Exercised call option {option_id}.")
            
        elif opt_type == 'P' and S0 < strike:
            self.settle_option(option_id, symbol, opt_type, quantity, strike, expiration, price)
            print(f"Exercised put option {option_id}.")
        
        # Remove the option from the portfolio, whether it was exercised or expired
        self.positions = self.positions[self.positions['option_id'] != option_id]

    def liquidate_portfolio(self):
        # Liquidate all option positions
        for idx, row in self.positions.iterrows():
            self.close_option(row['option_id'], row['underlying_symbol'], row['option_type'], row['quantity'], row['strike'], row['expiration'], row['unit_price'])
        
        # Liquidate all stock positions
        for idx, row in self.owned_stocks.iterrows():
            stock_quantity = row['quantity']
            stock_price = row['stock_price']
            if row['type'] == 'long':
                # Sell the long position
                self.money += stock_quantity * stock_price
            elif row['type'] == 'short':
                # Buy back the short position
                self.money -= stock_quantity * stock_price

class PurchasedOption:
    def __init__(self, optionid, symbol, opt_type, quantity, strike, expiration, price, stock_price):
        self.option_id = optionid
        self.underlying_symbol = symbol
        self.option_type = opt_type
        self.quantity = quantity
        self.strike = strike
        self.expiration = expiration
        self.price = price
        self.stock_price = stock_price


## 3. Backtesting 

In general: 

- Go through each row of the data, use the parameters
- Set current date to row['quote_date']
- Go through the data until the for loop hits a row['quote_date'] that is different than what it currently is defined (means that 1 day has passed) 


In [58]:
my_port = Portfolio(1_000_000)  # give myself 1 million dollars

current_day = None  # initialize the current day

temp = data[0:100]  # subset of data for backtesting
for index, row in temp.iterrows():
    if current_day is None:
        current_day = row['quote_date']  # Set the start date

    # calculate time to expiration in years
    T = (pd.to_datetime(row['expiration']) - pd.to_datetime(row['quote_date'])).days / 365.0
    
    # perform Monte Carlo simulation
    MC_option_price, MC_error = my_port.perform_monte_carlo_simulation(
        S0=row['underlying_ask_eod'],
        K=row['strike'],
        T=T,
        r=row['interest_rate'],
        sigma=row['implied_volatility_1545']
    )

    #    buy underpriced call options
    if row['option_type'] == 'C' and row['bid_eod'] < MC_option_price - MC_error:
        my_port.buy_option(
            data=row, 
            optionid=row['option_id'], 
            symbol=row['underlying_symbol'],
            opt_type=row['option_type'],
            quantity=row['bid_size_eod'],
            strike=row['strike'],
            expiration=row['expiration'],
            price=row['bid_eod'],
            stock_price=row['underlying_ask_eod']
        )

    # sell overpriced put options 
    elif row['option_type'] == 'P' and row['bid_eod'] > MC_option_price + MC_error:
        my_port.buy_option(
            data=row,
            optionid=row['option_id'], 
            symbol=row['underlying_symbol'],
            opt_type=row['option_type'],
            quantity=row['bid_size_eod'],
            strike=row['strike'],
            expiration=row['expiration'],
            price=row['bid_eod'],
            stock_price=row['underlying_ask_eod']
        )

    # Check if the day has ended
    if index < len(temp) - 1 and current_day != temp.iloc[index + 1]['quote_date']:
        # Process end-of-day option exercises
        for idx, position_row in my_port.positions.iterrows():
            current_market_price = position_row['stock_price']  # Use correct reference to the stock price

            # Exercise call options if they are in the money
            if position_row['option_type'] == 'C' and current_market_price > position_row['strike']:
                my_port.settle_option(
                    option_id=position_row['option_id'], 
                    symbol=position_row['underlying_symbol'], 
                    opt_type=position_row['option_type'], 
                    quantity=position_row['quantity'], 
                    strike=position_row['strike'], 
                    expiration=position_row['expiration'], 
                    price=position_row['unit_price']
                )

            # Exercise put options if they are in the money
            elif position_row['option_type'] == 'P' and current_market_price < position_row['strike']:
                my_port.settle_option(
                    option_id=position_row['option_id'], 
                    symbol=position_row['underlying_symbol'], 
                    opt_type=position_row['option_type'], 
                    quantity=position_row['quantity'], 
                    strike=position_row['strike'], 
                    expiration=position_row['expiration'], 
                    price=position_row['unit_price']
                )

        # Log the P&L for the day
        my_port.log_daily_pnl(current_day)

        # Update the current day
        current_day = temp.iloc[index + 1]['quote_date']

# Final liquidation of portfolio at the end of the backtest
my_port.liquidate_portfolio()
my_port.log_daily_pnl(current_day)  # Log the final P&L


Exercised call option NVDA_2023-01-06_65.0_C.
Exercised call option NVDA_2023-01-06_70.0_C.
Exercised call option NVDA_2023-01-06_75.0_C.
Exercised call option NVDA_2023-01-06_80.0_C.
Exercised call option NVDA_2023-01-06_85.0_C.
Exercised call option NVDA_2023-01-06_90.0_C.
Exercised call option NVDA_2023-01-06_95.0_C.
Exercised call option NVDA_2023-01-06_100.0_C.
Exercised call option NVDA_2023-01-06_105.0_C.
Exercised call option NVDA_2023-01-06_110.0_C.
Exercised call option NVDA_2023-01-06_111.0_C.
Exercised call option NVDA_2023-01-06_112.0_C.
Exercised call option NVDA_2023-01-06_113.0_C.
Exercised call option NVDA_2023-01-06_114.0_C.
Exercised call option NVDA_2023-01-06_115.0_C.
Exercised call option NVDA_2023-01-06_116.0_C.
Exercised call option NVDA_2023-01-06_117.0_C.
Exercised call option NVDA_2023-01-06_118.0_C.
Exercised call option NVDA_2023-01-06_119.0_C.
Exercised call option NVDA_2023-01-06_120.0_C.
Exercised call option NVDA_2023-01-06_121.0_C.
Exercised call optio

This is going to take forever if we run through all the options from January to December so we try the first 100.

In [59]:
# Extract the transaction data
transaction_data = my_port.transaction_log
# Filter for 'SELL' actions, which indicate option exercises
sell_transactions = transaction_data[transaction_data['action'] == 'SELL']
profit_loss_data = sell_transactions[['date', 'option_id', 'underlying_symbol', 'option_type', 'quantity', 'price', 'total_cost']]
# calculate P&L for each transaction
profit_loss_data['profit_loss'] = profit_loss_data['total_cost'] - (profit_loss_data['price'] * profit_loss_data['quantity'])

profit_loss_data['date'] = pd.to_datetime(profit_loss_data['date']).dt.date
profit_loss_data = profit_loss_data[['date', 'option_id', 'underlying_symbol', 'option_type', 'profit_loss']]
profit_loss_data = profit_loss_data.sort_values(by='date').reset_index(drop=True)
profit_loss_data


date                option_id underlying_symbol option_type  \
0   2023-01-06   NVDA_2023-01-06_65.0_C              NVDA           C   
1   2023-01-06  NVDA_2023-01-06_127.0_C              NVDA           C   
2   2023-01-06  NVDA_2023-01-06_128.0_C              NVDA           C   
3   2023-01-06  NVDA_2023-01-06_129.0_C              NVDA           C   
4   2023-01-06  NVDA_2023-01-06_130.0_C              NVDA           C   
5   2023-01-06  NVDA_2023-01-06_131.0_C              NVDA           C   
6   2023-01-06  NVDA_2023-01-06_132.0_C              NVDA           C   
7   2023-01-06  NVDA_2023-01-06_133.0_C              NVDA           C   
8   2023-01-06  NVDA_2023-01-06_134.0_C              NVDA           C   
9   2023-01-06  NVDA_2023-01-06_135.0_C              NVDA           C   
10  2023-01-06  NVDA_2023-01-06_126.0_C              NVDA           C   
11  2023-01-06  NVDA_2023-01-06_136.0_C              NVDA           C   
12  2023-01-06  NVDA_2023-01-06_138.0_C              NVDA           C   
13  2023-01-06  NVDA_2023-01-06_139.0_C              NVDA           C   
14  2023-01-06  NVDA_2023-01-06_140.0_C              NVDA           C   
15  2023-01-06  NVDA_2023-01-06_141.0_C              NVDA           C   
16  2023-01-06  NVDA_2023-01-06_142.0_C              NVDA           C   
17  2023-01-06  NVDA_2023-01-06_143.0_C              NVDA           C   
18  2023-01-06  NVDA_2023-01-06_146.0_P              NVDA           P   
19  2023-01-06  NVDA_2023-01-06_147.0_P              NVDA           P   
20  2023-01-06  NVDA_2023-01-06_148.0_P              NVDA           P   
21  2023-01-06  NVDA_2023-01-06_137.0_C              NVDA           C   
22  2023-01-06  NVDA_2023-01-06_125.0_C              NVDA           C   
23  2023-01-06  NVDA_2023-01-06_124.0_C              NVDA           C   
24  2023-01-06  NVDA_2023-01-06_123.0_C              NVDA           C   
25  2023-01-06   NVDA_2023-01-06_70.0_C              NVDA           C   
26  2023-01-06   NVDA_2023-01-06_75.0_C              NVDA           C   
27  2023-01-06   NVDA_2023-01-06_80.0_C              NVDA           C   
28  2023-01-06   NVDA_2023-01-06_85.0_C              NVDA           C   
29  2023-01-06   NVDA_2023-01-06_90.0_C              NVDA           C   
30  2023-01-06   NVDA_2023-01-06_95.0_C              NVDA           C   
31  2023-01-06  NVDA_2023-01-06_100.0_C              NVDA           C   
32  2023-01-06  NVDA_2023-01-06_105.0_C              NVDA           C   
33  2023-01-06  NVDA_2023-01-06_110.0_C              NVDA           C   
34  2023-01-06  NVDA_2023-01-06_111.0_C              NVDA           C   
35  2023-01-06  NVDA_2023-01-06_112.0_C              NVDA           C   
36  2023-01-06  NVDA_2023-01-06_113.0_C              NVDA           C   
37  2023-01-06  NVDA_2023-01-06_114.0_C              NVDA           C   
38  2023-01-06  NVDA_2023-01-06_115.0_C              NVDA           C   
39  2023-01-06  NVDA_2023-01-06_116.0_C              NVDA           C   
40  2023-01-06  NVDA_2023-01-06_117.0_C              NVDA           C   
41  2023-01-06  NVDA_2023-01-06_118.0_C              NVDA           C   
42  2023-01-06  NVDA_2023-01-06_119.0_C              NVDA           C   
43  2023-01-06  NVDA_2023-01-06_120.0_C              NVDA           C   
44  2023-01-06  NVDA_2023-01-06_121.0_C              NVDA           C   
45  2023-01-06  NVDA_2023-01-06_122.0_C              NVDA           C   
46  2023-01-06  NVDA_2023-01-06_149.0_P              NVDA           P   
47  2023-01-06  NVDA_2023-01-06_150.0_P              NVDA           P   

   profit_loss  
0        20.25  
1         3.41  
2         4.32  
3         4.95  
4         -3.5  
5        -1.92  
6         -8.4  
7       -74.82  
8       -18.92  
9       -48.38  
10       83.62  
11      -13.86  
12       -23.1  
13      -29.44  
14      -44.94  
15       -38.1  
16     -111.72  
17       -51.6  
18       -43.2  
19      -32.58  
20      -40.88  
21       -3.87  
22        7.35  
23      100.27  
24     

In [60]:
transaction_data

date action                option_id underlying_symbol option_type  \
0   2023-01-03    BUY   NVDA_2023-01-06_65.0_C              NVDA           C   
1   2023-01-03    BUY   NVDA_2023-01-06_70.0_C              NVDA           C   
2   2023-01-03    BUY   NVDA_2023-01-06_75.0_C              NVDA           C   
3   2023-01-03    BUY   NVDA_2023-01-06_80.0_C              NVDA           C   
4   2023-01-03    BUY   NVDA_2023-01-06_85.0_C              NVDA           C   
..         ...    ...                      ...               ...         ...   
98  2023-01-06   SELL  NVDA_2023-01-06_146.0_P              NVDA           P   
99  2023-01-06   SELL  NVDA_2023-01-06_147.0_P              NVDA           P   
100 2023-01-06   SELL  NVDA_2023-01-06_148.0_P              NVDA           P   
101 2023-01-06   SELL  NVDA_2023-01-06_149.0_P              NVDA           P   
102 2023-01-06   SELL  NVDA_2023-01-06_150.0_P              NVDA           P   

    quantity  price  total_cost  
0         25  77.35     1933.75  
1         25  72.35     1808.75  
2         25  67.35     1683.75  
3         35  61.20     2142.00  
4         37  57.15     2114.55  
..       ...    ...         ...  
98        20   5.00       56.80  
99        18   5.65       69.12  
100       28   6.30      135.52  
101       23   7.05      134.32  
102       24   7.80      164.16  

[103 rows x 8 columns]

In [61]:
my_port.money

1000157.7900000002

As we can see, we end up with a positive return from our simulation, resulting in a final portfolio value of \\$1,000,157.79 from an initial investment of \\$1,000,000. This outcome demonstrates that the strategy was successful in identifying and capitalizing on mispriced options in the market. The application of Monte Carlo simulations and binomial option pricing models provided a robust framework for decision-making, allowing for careful consideration of when to exercise options and when to hold them.

